nscplugin/src/main/scala-2/scala/scalanative/nscplugin/NirGenExpr.scala

```diff
@@ -617,24 +617,24 @@ trait NirGenExpr[G <: nsc.Global with Singleton] { self: NirGenPhase[G] =>
         buf.extract(qual, Seq(index), unwind)
       } else {
         val ty = genType(tree.symbol.tpe)
-        val qual = genExpr(qualp)
         val name = genFieldName(tree.symbol)
         if (sym.owner.isExternModule) {
           val externTy = genExternType(tree.symbol.tpe)
           genLoadExtern(ty, externTy, tree.symbol)
         } else {
+          val qual = genExpr(qualp)
           buf.fieldload(ty, qual, name, unwind)
         }
       }
     }
 
-    def genStaticMember(sym: Symbol)(implicit pos: nir.Position): Val = {
+    def genStaticMember(receiver: Tree, sym: Symbol)(implicit
+        pos: nir.Position
+    ): Val = {
       if (sym == BoxedUnit_UNIT) {
         Val.Unit
       } else {
-        val ty = genType(sym.tpe)
-        val module = genModule(sym.owner)
-        genApplyMethod(sym, statically = true, module, Seq())
+        genApplyStaticMethod(sym, receiver, Seq())
```

Comment:
This change is in fact not backward binary compatible. It can break the scenario where there exists a third-party library implementing some part of the JDK (such as scala-java-time) and compiled with 0.4.1. That library would not have static forwarders for the members. But now the compiler will emit calls to those non-existent static forwarders, which will break.

I am not actually aware of any third-party JDK implementation for Scala Native. So if none exists, we're good.

If there is, we can work around this problem in two possible ways:
* Either we introduce some linking magic for compatibility: when trying to reach a static method from a class compiled with 0.4.0 or 0.4.1, if it does not exist, we try calling a method of the same name in the companion object instead. This is completely doable (and we've done worse things in the history of Scala.js IR), but non-trivial.
* Or we still capitalize on the not-so-mature aspect of Scala Native to say that it is a reasonable breakage: there can only be 1-3 such libraries in the wild, and they would be able to recompile and republish newer versions with Scala Native 0.4.2. Those newer versions would not break their own backward binary compatibility, so we don't create a fracture in the binary ecosystem. The only downside is that users of those libraries would have to upgrade those libraries at the same time as they upgrade Scala Native. But this is probably a reasonable price to pay given the state of the ecosystem.

Reply:
I've added a step in `Reach` that does rewrite calls to static methods. I've tested it locally by rewriting all static calls (by disabling one of the conditions) - there were 2 corner cases that I've addressed: 
 * ImplClasses was containing only static methods, I've decided to not produce static methods at all for all ImplClasses members 
 * `lifterTree` (Scala intrinsic method?) was defined only as static in class - I've added additional case that would try to use methods defined in class instead of companion module

Reply:
> * `lifterTree` (Scala intrinsic method?) was defined only as static in class - I've added additional case that would try to use methods defined in class instead of companion module

I think this was an artifact of how you stress-tested the rewrites. I don't think this can happen in practice, because `lifterTree` should only be called from the same compilation unit, I think (so separate compilation and binary compat are not a concern).

Perhaps a more accurate stress-test would be to disable the generation of all *static forwarders*, but otherwise keep everything the same (and in particular, emit already-static methods normally, as static methods).

Reply:
After disabling static forwarders everything seems to be working fine, all unit tests in Scala 2.11 2.12 and 2.13 are able to link and are passing the tests.
In seed `liftedTree` methods are defined as `private static` so they should never be a problem for us.